sympy/solvers/solveset.py

```diff
@@ -2594,14 +2674,25 @@ def linear_eq_to_matrix(equations, *symbols):
             Eq or Matrix.
             '''))
 
-    A, b = [], []
-    for i, f in enumerate(equations):
-        if isinstance(f, Equality):
-            f = f.rewrite(Add, evaluate=False)
-        coeff_list = linear_coeffs(f, *symbols)
-        b.append(-coeff_list.pop())
-        A.append(coeff_list)
-    A, b = map(Matrix, (A, b))
+    # convert Eq now because _linear_eq_to_dict will allow cancellation
+    # of nonlinear terms -- this can be removed when the XFAIL for this
+    # passes
+    equations = [i.rewrite(Add, evaluate=False) if isinstance(i, Eq)
+        else i for i in equations]
+
+    # construct the dictionaries
+    try:
+        eq, c = _linear_eq_to_dict(equations, symbols, strict=strict)
+    except PolyNonlinearError as err:
+        raise NonlinearError(str(err)) from err
+    ix = dict(zip(symbols, range(len(symbols))))
+    A = zeros(len(eq), len(symbols))
+    b = zeros(len(eq), 1)
+    for i, d in enumerate(eq):
+        for s in d:
+            j = ix[s]
+            A[i, j] = d[s]
+        b[i, 0] = -c[i]
```

Comment:
The most efficient way to do this is the way that `sympy_dict_to_dm` does it. We just need to make sure that it chooses the right domain (ZZ, QQ or EXRAW) because that's how the internals of Matrix works:
https://github.com/sympy/sympy/blob/3c6e48ea7912fba507d837d7f4f3395a9bba238f/sympy/matrices/repmatrix.py#L30-L45
The `_dod_to_domain_matrix` method basically does what is needed:
https://github.com/sympy/sympy/blob/3c6e48ea7912fba507d837d7f4f3395a9bba238f/sympy/matrices/repmatrix.py#L102
Perhaps there should be a public Matrix method like `from_dod`  since that is the most efficient way to construct a matrix.

Reply:
Yes, we have `Matrix.todod`. Let's let that be another PR.

Reply:
SparseMatrix now knows how to parse dod.